src/Kestrel.Core/Internal/Http/Http1Connection.cs

```diff
@@ -152,64 +155,88 @@ public void ParseRequest(ReadOnlySequence<byte> buffer, out SequencePosition con
+
+            //Done:;
         }
 
-        public bool TakeStartLine(ReadOnlySequence<byte> buffer, out SequencePosition consumed, out SequencePosition examined)
+        public bool TakeStartLine(ref BufferReader<byte> reader, long length)
         {
-            var overLength = false;
-            if (buffer.Length >= ServerOptions.Limits.MaxRequestLineSize)
+            bool result;
+            if (length >= ServerOptions.Limits.MaxRequestLineSize)
             {
-                buffer = buffer.Slice(buffer.Start, ServerOptions.Limits.MaxRequestLineSize);
-                overLength = true;
+                result = TakeOverlengthStartLine(ref reader);
             }
-
-            var result = _parser.ParseRequestLine(new Http1ParsingHandler(this), buffer, out consumed, out examined);
-            if (!result && overLength)
+            else
             {
-                BadHttpRequestException.Throw(RequestRejectionReason.RequestLineTooLong);
+                result = _parser.ParseRequestLine(new Http1ParsingHandler(this), ref reader);
             }
 
             return result;
         }
 
-        public bool TakeMessageHeaders(ReadOnlySequence<byte> buffer, out SequencePosition consumed, out SequencePosition examined)
+        private bool TakeOverlengthStartLine(ref BufferReader<byte> reader)
         {
-            // Make sure the buffer is limited
-            bool overLength = false;
-            if (buffer.Length >= _remainingRequestHeadersBytesAllowed)
-            {
-                buffer = buffer.Slice(buffer.Start, _remainingRequestHeadersBytesAllowed);
+            // More data in the reader than we want to allow, cap it
+            var sequence = reader.Sequence;
+            reader = new BufferReader<byte>(sequence.Slice(sequence.Start, ServerOptions.Limits.MaxRequestLineSize));
+
+            var result = _parser.ParseRequestLine(new Http1ParsingHandler(this), ref reader);
 
-                // If we sliced it means the current buffer bigger than what we're
-                // allowed to look at
-                overLength = true;
+            if (result)
+            {
+                // Need to reset the reader to the right end point
+                var consumed = reader.Consumed;
+                reader = new BufferReader<byte>(sequence);
+                reader.Advance(consumed);
+            }
+            else
+            {
+                BadHttpRequestException.Throw(RequestRejectionReason.RequestLineTooLong);
             }
 
-            var result = _parser.ParseHeaders(new Http1ParsingHandler(this), buffer, out consumed, out examined, out var consumedBytes);
-            _remainingRequestHeadersBytesAllowed -= consumedBytes;
+            return result;
+        }
 
-            if (!result && overLength)
+        public bool TakeMessageHeaders(ref BufferReader<byte> reader, long length)
+        {
+            var consumed = reader.Consumed;
+            var remaining = length - consumed;
```

Comment:
This annoying extra bookkeeping to keep track of the underlying ReadOnlySequence's length makes me think that BufferReader should offer a cached, lazily-initialized Length property. An efficient Remaining property could also be built on top of this.

Reply:
It is definitely worth considering. I'm being very cautious about growing the size of the reader, but at first blush it seems useful enough to take the extra field.